include/operations/extension_trees.h

```diff
@@ -0,0 +1,103 @@
+/***************************************************************************
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename extension_trees.h
+ *
+ **************************************************************************/
+
+#ifndef SYCL_BLAS_EXTENSION_TREES_H
+#define SYCL_BLAS_EXTENSION_TREES_H
+
+#include <CL/sycl.hpp>
+
+namespace blas {
+
+/*!
+ * @brief Determines which type of reduction to perform
+ */
+enum class Reduction_t : int {
+  full = 0,  // Not implemented yet
+  partial_rows = 1,
+  partial_columns = 2  // Not implemented yet
+};
+
+/*!
+ * @brief Wrapper around the reduction.
+ *
+ * The executor will specialize the execution for every reduction type and use
+ * the specific reduction classes
+ */
+template <typename operator_t, typename input_t, typename output_t, int ClSize,
+          int WgSize, int WorkPerItem, typename element_t, int Reduction_type>
+class Reduction {
+ public:
+  using index_t = typename std::make_signed<typename input_t::index_t>::type;
```

Comment:
It is something that I saw in gemm implementation, a signed index type is always used. I assume it is to avoid problems with all the `int` to `index_t` conversions (the template parameters are `int`)
@mehdi-goli any opinion on this?

Reply:
It has been changed in GEMM to signed due to the way that GEMM algorithm has been implemented. In Gemm if the index goes negative, it does not load the data. If you have such a concept in your kernel then apply the make_signed, if not, it is not necessary to be here.

Reply:
Ok, I don't need it so I removed the `make_signed`